lib/rubygems/installer.rb

```diff
@@ -256,6 +260,8 @@ def install
       write_cache_file
     end
 
+    File.chmod(dir_mode, *Dir.glob(gem_dir+"/**/").map {|path| path.untaint}) if dir_mode
```

Comment:
That code will also blow up if the list is too long due to the giant splat. Code should iterate over the list and call chmod for each entry (also faster because no extra Array needed)